sdk/cdsclient/client_workflow.go

```diff
@@ -253,7 +266,107 @@ func (c *client) WorkflowCachePush(projectKey, ref string, tarContent io.Reader)
 	return nil
 }
 
+func (c *client) workflowCachePushIndirectUpload(projectKey, ref string, tarContent io.Reader) error {
+	url := fmt.Sprintf("/project/%s/cache/%s/url", projectKey, ref)
+
+	cacheObj := sdk.Cache{}
+	code, err := c.PostJSON(url, nil, &cacheObj)
+	if err != nil {
+		return err
+	}
+
+	if code >= 400 {
+		return fmt.Errorf("HTTP Code %d", code)
+	}
+
+	if err := c.workflowCachePushIndirectUploadPost(cacheObj.TmpURL, tarContent); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *client) workflowCachePushIndirectUploadPost(url string, tarContent io.Reader) error {
+	//Post the file to the temporary URL
+	var retry = 10
+	var globalErr error
+	var body []byte
+	for i := 0; i < retry; i++ {
+		req, errRequest := http.NewRequest("PUT", url, tarContent)
+		if errRequest != nil {
+			return errRequest
+		}
+		req.Header.Set("Content-Type", "application/tar")
+
+		var resp *http.Response
+		resp, globalErr = http.DefaultClient.Do(req)
+		if globalErr == nil {
+			defer resp.Body.Close()
+
+			var err error
+			body, err = ioutil.ReadAll(resp.Body)
+			if err != nil {
+				globalErr = err
+				continue
+			}
+
+			if resp.StatusCode >= 300 {
+				globalErr = fmt.Errorf("[%d] Unable to upload cache: (HTTP %d) %s", i, resp.StatusCode, string(body))
+				continue
+			}
+
+			break
+		}
+		time.Sleep(500 * time.Millisecond)
+	}
+
+	return globalErr
+}
+
 func (c *client) WorkflowCachePull(projectKey, ref string) (io.Reader, error) {
+	store := new(sdk.ArtifactsStore)
+	_, _ = c.GetJSON("/artifact/store", store)
+	if store.TemporaryURLSupported {
+		return c.workflowCacheIndirectDownload(projectKey, ref)
+	}
+	return c.workflowCacheDirectDownload(projectKey, ref)
+}
+
+func (c *client) workflowCacheIndirectDownload(projectKey, ref string) (io.Reader, error) {
```

Comment:
i don't think we need 2 functions for direct and indirect.

df https://github.com/ovh/cds/blob/master/sdk/cdsclient/client_workflow.go#L131

Reply:
done